### ISSUE TITLE
Add CLAP (Contrastive Language-Audio Pretraining) model

### ISSUE THREAD

**Issue by yassienashrafwasfy** [keras-team/keras-hub#2739]
- CLAP is the audio equivalent of CLIP: joint audio-text embeddings via contrastive learning, enabling zero-shot audio classification and text-to-audio retrieval. The laion/clap-htsat-fused checkpoint has 20.5M downloads last month on HuggingFace.

- With Swin Transformer and RoBERTa already in KerasHub, most building blocks are there. Main work will be the weight conversion script and mirroring the CLIP design pattern internally.

- Model License: Apache 2.0.

- Planning to start June 20, would love a green light from the team before diving in.

- Note on BLIP-2: Postponing for now. OPT-2.7B and OPT-6.7B are done. FlanT5 needs architecture changes and I need more time to understand how InstructBLIP works before continuing.

Resources:

- [Paper](https://arxiv.org/pdf/2211.06687)
- [Model Card HF](https://huggingface.co/laion/clap-htsat-fused)

**Comment by amitsrivastava78** [comment]
Thanks for the heads up, Before you dive in, I just want to align on a few architectural details so we are completely clear on the scope.

First off, we won't be able to just reuse the existing Swin Transformer code for the audio side. Since HTSAT adapts the Swin architecture specifically for audio spectrograms, there are a few areas where we'll need to write fresh code:

1. The input pipeline: We're going from a raw waveform to a mel spectrogram, and then to 2D patch embedding. Since we aren't dealing with image pixels, we'll need audio-specific preprocessing (like window size, hop length, and n_mels) that doesn't exist in our current Swin backbone.

2. Feature fusion: To handle audio clips of varying lengths, the code needs to interpolate and concatenate features from both the full-length and chunk-level mel spectrograms. This is the main novel piece from the LAION-CLAP checkpoint, and it doesn't really have a parallel in Swin or CLIP.

3. Patch merging: The strides here need to be tuned specifically to the time-frequency domain rather than spatial image patches.

The good news is that for the text encoder, RoBERTa is already in KerasHub, so that should be a very clean reuse.

For the overall architecture, let's mirror the CLIPBackbone pattern exactly. We can build it as a main wrapper with contrastive projection heads, containing two sub-components: ClapAudioEncoder and ClapTextEncoder.

One major thing we need to decide on is the audio preprocessing dependency. Keras currently has keras.ops.fft, but mel spectrogram computation isn't built-in. Let me know if you plan to build the mel spectrogram computation directly as a custom Keras layer (which I strongly prefer, as it keeps us backend-agnostic) or if you want to require a librosa or torchaudio dependency at preprocessing time. We need to lock in a decision on this before the design is finalized.

To summarize, your main implementation chunks will be:

1. ClapAudioEncoder (the HTSAT backbone with the mel-spectrogram input layer and feature fusion).

2. ClapTextEncoder (a thin wrapper around our existing RoBERTa backbone).
 
3. ClapBackbone (following the CLIP pattern with projection heads).

4. A script to handle weight conversion from laion/clap-htsat-fused.

Happy to give the formal green light once you confirm the preprocessing approach! Just drop a quick reply confirming how you plan to tackle the feature fusion and the audio preprocessing layer, and you are good to start on June 20th.

**Comment by yassienashrafwasfy** [comment]
@amitsrivastava78 

Forgot to mention the audio converter, but the plan looks good. For the audio converter specifically, I'll replicate Whisper's Audio Converter from Keras Hub with some tweaks for our use case. I already ran into this exact converter issue during my CNN14 PANNs port and got the converter working with low error compared to the original, but the port itself was rejected unfortunately. So I have a solid head start on the converter side. For the feature fusion part, I'll look into it and get back to you.

**Comment by yassienashrafwasfy** [comment]
@amitsrivastava78 

Hey, sorry for the delay, life's been busy. Back on it now.

Spent time on the paper and the LAION codebase. Here's the fusion plan to match the HF checkpoint.

The `ClapAudioAFFBlock` in the checkpoint is plain AFF

Both branches run on `xa = x + residual`:

- `local_att`: Conv2D(1×1) → BN → ReLU → Conv2D(1×1) → BN. Keeps spatial dims, gates per channel per position.
- `global_att`: same stack with `AdaptiveAvgPool2d(1)` in front. Pools spatial first, gates per channel over the whole clip.

Then:

```python
wei = sigmoid(local_att(xa) + global_att(xa))
xo  = 2 * x * wei + 2 * residual * (1 - wei)
```
Mapping to the printout:

- `proj` (Conv2d 4×4): global (slot 0).
- `mel_conv2d` (Conv2d 4×12): local (slots 1-3).
- `fusion_model`: the AFF block, gates local against global.

### Resources

### Paper Review
[![Screenshot-2026-05-19-174142.png](https://i.postimg.cc/SR7yJRXn/Screenshot-2026-05-19-174142.png)](https://postimg.cc/T518sRJx)

### HF Checkpoint Summary

[![Screenshot-2026-05-19-170802.png](https://i.postimg.cc/K8GkbS15/Screenshot-2026-05-19-170802.png)](https://postimg.cc/ZCQKPQF9)

### Paper Details about Fusion Concerns

[![Screenshot-2026-05-19-171521.png](https://i.postimg.cc/jSZgXTVw/Screenshot-2026-05-19-171521.png)](https://postimg.cc/Hr89Yfkd)

### Src Code for Fusion

[Fusion Src Code](https://github.com/LAION-AI/CLAP/blob/main/src/laion_clap/clap_module/feature_fusion.py)

**Comment by amitsrivastava78** [comment]
Thanks — the AFF breakdown and slot mapping (proj → global, mel_conv2d → local, fusion_model → AFF) resolves the main open question. Plain-AFF (not iAFF) confirmation against the paper and LAION source is exactly what was needed.

A few small things to pin down before kickoff:

v1 scope — only laion/clap-htsat-fused, other LAION variants deferred?
Text encoder — reuse RobertaBackbone directly + projection head, or wrapper needed?
Numerical parity targets — please pin explicit thresholds (max-abs / cosine) for mel-spectrogram, audio embedding, text embedding, and joint logits up front.
Tests — standalone AFF block test (novel piece) + end-to-end zero-shot classification on a fixture clip vs HF outputs.

Happy to greenlight once these are clarified.

**Comment by yassienashrafwasfy** [comment]
@amitsrivastava78 

- Sorry for the late reply, senior year is no joke and life gets busy.

- Three weeks to graduation, after that I'm fully free. Sadly had to postpone until June 20th, but the plan is finishing BLIP-2 T5-Flan variants and instruct, then the remaining CLAP variants: htsat-unfused, larger-clap-music, larger-clap-general.

- On the RoBERTa question: I'll transfer weights directly from the CLAP checkpoint to randomly initialized Roberta. RoBERTa starts with masked language modeling but contrastive learning is what aligns the embeddings across both modalities.

- For numerical validation I'll check each block independently first. Already hit a low numerical difference on BLIP-2 OPT-2.7B so CLAP should be easier given its smaller size.
- For tests I'll look into AFF tests as I go, validate each block numerically first, then run a full pipeline comparison between the Keras and HF versions targeting 1e-4 to 1e-5 numerical divergence.

### Logs from the BLIP-2 OPT-2.7B conversion script (Previous Keras-Hub work)
[Kaggle Notebook](https://www.kaggle.com/code/yassienwasfy/blip-2-opt2-7-keras-conversion-script)

**Comment by yassienashrafwasfy** [comment]
hi @amitsrivastava78 

any `updates`?
